File: c#/#654_max_binary_tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public TreeNode ConstructMaximumBinaryTree(int[] nums)
    {
        return ConstructMaximumBinaryTree(nums, 0, nums.Length - 1);
    }
    
    //helper func
    private TreeNode ConstructMaximumBinaryTree(int[] nums, int left, int right)
    {
        if (left > right)
        {
            return null;
        }
        
        int maxIndex = FindMaxIndex(nums, left, right);
        // Create a new TreeNode with the maximum value as its value
        TreeNode root = new TreeNode(nums[maxIndex]);
        
        root.left = ConstructMaximumBinaryTree(nums, left, maxIndex - 1);
        root.right = ConstructMaximumBinaryTree(nums, maxIndex + 1, right);
        
        return root;
    }
    
    private int FindMaxIndex(int[] nums, int left, int right)
    {
        int maxIndex = left;
        for (int i = left + 1; i <= right; ++i)
        {
            if (nums[i] > nums[maxIndex])
            {
                maxIndex = i;
            }
        }
    
        return maxIndex;
    }
}

File: cpp/#654_max_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums)
    {
        return ConstructMaximumBinaryTree(nums, 0, nums.size() - 1);
    }

private:
    TreeNode* ConstructMaximumBinaryTree(vector<int>& nums, int left, int right)
    {
        if (left > right)
        {
            return nullptr;
        }

        int maxIndex = FindMaxIndex(nums, left, right);
        TreeNode* root = new TreeNode(nums[maxIndex]);

        root->left = ConstructMaximumBinaryTree(nums, left, maxIndex - 1);
        root->right = ConstructMaximumBinaryTree(nums, maxIndex + 1, right);

        return root;
    }

    int FindMaxIndex(vector<int>& nums, int left, int right)
    {
        int maxIndex = left;
        for (int i = left + 1; i <= right; ++i)
        {
            if (nums[i] > nums[maxIndex])
            {
                maxIndex = i;
            }
        }

        return maxIndex;
    }
};